In [150]:
import pandas as pd

In [151]:
df = pd.read_csv('C:/Users/USERAS/Desktop/Cit/KNN (K-Nearest Neighbors)/KNNAlgorithmDataset.csv')
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [152]:
df.isna().sum()

id                           0
diagnosis                    0
radius_mean                  0
texture_mean                 0
perimeter_mean               0
area_mean                    0
smoothness_mean              0
compactness_mean             0
concavity_mean               0
concave points_mean          0
symmetry_mean                0
fractal_dimension_mean       0
radius_se                    0
texture_se                   0
perimeter_se                 0
area_se                      0
smoothness_se                0
compactness_se               0
concavity_se                 0
concave points_se            0
symmetry_se                  0
fractal_dimension_se         0
radius_worst                 0
texture_worst                0
perimeter_worst              0
area_worst                   0
smoothness_worst             0
compactness_worst            0
concavity_worst              0
concave points_worst         0
symmetry_worst               0
fractal_dimension_worst      0
Unnamed:

In [153]:
df.shape

(569, 33)

In [154]:
df = df.drop('Unnamed: 32' , axis=1)   # For KNN regression Have to remove String related and Nan columns or can handle those using labeling  else     eror:  ValueError: could not convert string to float: 'B'
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

In [155]:
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor

In [156]:
df.dtypes

id                           int64
diagnosis                   object
radius_mean                float64
texture_mean               float64
perimeter_mean             float64
area_mean                  float64
smoothness_mean            float64
compactness_mean           float64
concavity_mean             float64
concave points_mean        float64
symmetry_mean              float64
fractal_dimension_mean     float64
radius_se                  float64
texture_se                 float64
perimeter_se               float64
area_se                    float64
smoothness_se              float64
compactness_se             float64
concavity_se               float64
concave points_se          float64
symmetry_se                float64
fractal_dimension_se       float64
radius_worst               float64
texture_worst              float64
perimeter_worst            float64
area_worst                 float64
smoothness_worst           float64
compactness_worst          float64
concavity_worst     

In [157]:
# Handling String related column else value error coz regression can use only numerical values

# Drop the target column first (area_mean)
X = df.drop('area_mean', axis=1)  # Features[independent variables] (everything except 'area_mean')

X = pd.get_dummies(X)             # One hot encoding - | 1 | 0 | 0 | , Coz 'diagnosis'  column has string values . One hot will convert them in numeric values

y = df['area_mean']               # Labels (dependent variable or target)


# Split into train and test sets
xtrain, xtest, ytrain, ytest = train_test_split(X, y, test_size=0.2, random_state= 42)

#test_size=0.2	The proportion of the data to reserve for testing. 0.2 = 20% test, 80% train.
#random_state=42	Sets a fixed random seed so that the split is reproducible. Use any integer. Same seed = same result every time.
#Without random_state, the split will be different every time you run it. 
#With random_state=42 (or any number), the split will be exactly the same every time. random_state is used to make sure the data is split the same way every time you run the code.
#Why 42? It's just a popular, well-known default (from the book "The Hitchhiker's Guide to the Galaxy", where 42 is the "Answer to the Ultimate Question").

In [158]:
knn = KNeighborsRegressor(n_neighbors= 3)  # for regression 5% of the data at most

In [159]:
knn.fit(xtrain,ytrain)

KNeighborsRegressor(n_neighbors=3)

In [160]:
knn.predict(xtest)

array([ 553.3       ,  605.5       ,  570.26666667,  564.1       ,
        498.46666667,  974.        , 1388.66666667,  755.66666667,
        615.36666667,  510.06666667,  497.5       ,  942.5       ,
        754.23333333,  834.4       , 1256.6       ,  615.36666667,
        702.6       ,  506.06666667,  389.86666667,  405.73333333,
        380.2       ,  362.76666667,  601.1       ,  531.3       ,
        457.26666667,  750.1       ,  483.53333333,  506.06666667,
        516.73333333,  605.5       ,  377.46666667,  363.03333333,
        439.53333333, 1256.6       ,  558.86666667,  410.8       ,
        808.5       ,  549.56666667,  774.3       ,  643.93333333,
        389.83333333,  412.93333333,  482.03333333,  638.36666667,
        343.8       ,  334.56666667,  426.7       ,  553.66666667,
       1256.6       ,  627.16666667,  879.83333333,  996.13333333,
        638.36666667,  436.46666667,  447.4       , 1281.53333333,
        394.4       ,  615.36666667,  595.2       ,  422.46666

In [161]:
from sklearn.metrics import mean_squared_error, r2_score

ypred = knn.predict(xtest)
print("MSE:", mean_squared_error(ytest, ypred))
print("R² Score:", r2_score(ytest, ypred))
# Conduct the same operation but with linear regression and see the difference

MSE: 99174.60390838208
R² Score: 0.13994146202086233


# Interpretation : 

The Mean Squared Error (MSE) of about 99,175 means that, on average, the squared difference between your model’s predictions and the actual values is quite large. The R² score of approximately 0.14 indicates that the model explains only about 14% of the variation in the target variable, which is quite low. This suggests that the model isn’t capturing the data patterns very well and has limited predictive power on this dataset.

| R² Score Range | Interpretation                   |
| -------------- | -------------------------------- |
| 0.9 to 1.0     | Excellent fit                    |
| 0.7 to 0.9     | Good fit                         |
| 0.5 to 0.7     | Moderate/acceptable fit          |
| 0 to 0.5       | Poor fit                         |
| < 0            | Very poor; model worse than mean |


In [162]:
# The code finds the best number of neighbors for your KNN model automatically, then evaluates how well it predicts unseen data

from sklearn.model_selection import  GridSearchCV
from sklearn.metrics import mean_squared_error
import numpy as np



# Define parameter grid for neighbors
param_grid = {'n_neighbors': range(1, 51)}    # Tries different values for n_neighbors from 1 to 50.

# Perform grid search with 5-fold CV using neg MSE as scoring
grid_search = GridSearchCV(knn, param_grid, cv=5, scoring='neg_mean_squared_error')   # GridSearchCV helps automatically find the best hyperparameters.
grid_search.fit(xtrain, ytrain)

# Get best number of neighbors
best_k = grid_search.best_params_['n_neighbors']    # Finds which n_neighbors gave the lowest error.
print("Best n_neighbors:", best_k)

# Predict using the best model
ypred = grid_search.best_estimator_.predict(xtest)  # Uses the best found model to predict the target for test data.

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(ytest, ypred))    # mean_squared_error measures how bad the model predictions are.
print(f"RMSE: {rmse:.2f}")

Best n_neighbors: 2
RMSE: 302.42


# Interpretation :
The best number of neighbors for your KNN model is 2, meaning the model predicts the target value by looking at the 2 closest points in the training data. Using this setting, your model’s predictions on new, unseen data have an average error (RMSE) of about 302.42, which means the predicted values differ from the actual values by around 302 units on average. Whether this error is acceptable depends on the scale of the data, but overall, the model is tuned to balance fitting the data well without overfitting or underfitting.